File: packages/native/android/src/main/java/com/harshy/sdk/Road.kt
package com.harshy.sdk

import kotlin.math.abs
import kotlin.math.sqrt

internal const val ROAD_WINDOW_MS = 1000.0

internal fun verticalLinearAccel(sample: ImuSample): Double {
  val linear = sample.linearAccel
  val accel = linear ?: sample.accel
  val gravity = sample.gravity
  if (gravity != null) {
    val mag = magnitude(gravity)
    if (mag > 0.5) {
      val projected = (accel.x * gravity.x + accel.y * gravity.y + accel.z * gravity.z) / mag
      return abs(if (linear != null) projected else projected - mag)
    }
  }
  return abs(if (linear != null) linear.z else sample.accel.z)
}

internal fun assessRoad(
  location: List<LocationSample>,
  imu: List<ImuSample>,
  startedAtMs: Double,
  jerkSettleMs: Double,
  minSpeedMps: Double,
  windowMs: Double = ROAD_WINDOW_MS,
): List<LocationSample> {
  val settleUntil = startedAtMs + jerkSettleMs
  if (imu.isEmpty()) {
    return location.map { sample ->
      sample.copy(roadRmsMps2 = sample.roadRmsMps2)
    }
  }

  val samples = imu.sortedBy { it.t }
  var start = 0

  return location.map { sample ->
    if (sample.roadRmsMps2 != null) {
      return@map sample
    }
    val speed = sample.speedMps ?: 0.0
    if (sample.t < settleUntil || speed < minSpeedMps) {
      return@map sample.copy(roadRmsMps2 = null)
    }

    val lo = sample.t - windowMs
    val hi = sample.t
    while (start < samples.size && samples[start].t < lo) {
      start += 1
    }

    var sumSq = 0.0
    var n = 0
    for (i in start until samples.size) {
      val imuSample = samples[i]
      if (imuSample.t > hi) {
        break
      }
      if (imuSample.t < settleUntil) {
        continue
      }
      val vertical = verticalLinearAccel(imuSample)
      sumSq += vertical * vertical
      n += 1
    }

    sample.copy(roadRmsMps2 = if (n == 0) null else sqrt(sumSq / n))
  }
}

File: packages/core/src/road.ts
import { magnitude } from "./geo.js";
import type { ImuSample, LocationSample } from "./types.js";

/** Vertical linear-accel RMS window assigned to each GPS sample. */
export const ROAD_WINDOW_MS = 1000;

export type AssessRoadOptions = {
  startedAtMs: number;
  jerkSettleMs: number;
  minSpeedMps: number;
  windowMs?: number;
};

/**
 * World-up linear acceleration (m/s²). Phone-Z is not “up” in a cup holder —
 * project onto the gravity vector when present.
 */
export function verticalLinearAccel(sample: ImuSample): number {
  const linear = sample.linearAccel;
  const accel = linear ?? sample.accel;
  const gravity = sample.gravity;
  if (gravity) {
    const mag = magnitude(gravity);
    if (mag > 0.5) {
      const projected =
        (accel.x * gravity.x + accel.y * gravity.y + accel.z * gravity.z) / mag;
      return Math.abs(linear ? projected : projected - mag);
    }
  }
  return Math.abs(linear ? linear.z : accel.z);
}

/**
 * Pavement roughness along the GPS track. Does not emit events or change score.
 * Idle / settle windows are `null`. When IMU is empty, existing `roadRmsMps2` is kept
 * so compact lab JSON (IMU dropped) still retunes without wiping the path measure.
 */
export function roadRmsForLocation(
  sample: LocationSample,
  imu: readonly ImuSample[],
  options: AssessRoadOptions,
): number | null {
  const windowMs = options.windowMs ?? ROAD_WINDOW_MS;
  const settleUntil = options.startedAtMs + options.jerkSettleMs;
  const speed = sample.speedMps ?? 0;
  if (sample.t < settleUntil || speed < options.minSpeedMps || imu.length === 0) {
    return null;
  }

  const lo = sample.t - windowMs;
  const hi = sample.t;
  let sumSq = 0;
  let n = 0;
  for (let i = imu.length - 1; i >= 0; i -= 1) {
    const imuSample = imu[i];
    if (!imuSample) {
      continue;
    }
    if (imuSample.t > hi) {
      continue;
    }
    if (imuSample.t < lo) {
      break;
    }
    if (imuSample.t < settleUntil) {
      continue;
    }
    const vertical = verticalLinearAccel(imuSample);
    sumSq += vertical * vertical;
    n += 1;
  }
  return n === 0 ? null : Math.sqrt(sumSq / n);
}

export function assessRoad(
  location: readonly LocationSample[],
  imu: readonly ImuSample[],
  options: AssessRoadOptions,
): LocationSample[] {
  const windowMs = options.windowMs ?? ROAD_WINDOW_MS;
  const settleUntil = options.startedAtMs + options.jerkSettleMs;

  if (imu.length === 0) {
    return location.map((sample) => ({
      ...sample,
      roadRmsMps2: sample.roadRmsMps2 ?? null,
    }));
  }

  const samples = [...imu].sort((a, b) => a.t - b.t);
  let start = 0;

  return location.map((sample) => {
    if (sample.roadRmsMps2 != null) {
      return sample;
    }
    const speed = sample.speedMps ?? 0;
    if (sample.t < settleUntil || speed < options.minSpeedMps) {
      return { ...sample, roadRmsMps2: null };
    }

    const lo = sample.t - windowMs;
    const hi = sample.t;
    while (start < samples.length && (samples[start]?.t ?? 0) < lo) {
      start += 1;
    }

    let sumSq = 0;
    let n = 0;
    for (let i = start; i < samples.length; i += 1) {
      const imuSample = samples[i];
      if (!imuSample || imuSample.t > hi) {
        break;
      }
      if (imuSample.t < settleUntil) {
        continue;
      }
      const vertical = verticalLinearAccel(imuSample);
      sumSq += vertical * vertical;
      n += 1;
    }

    return {
      ...sample,
      roadRmsMps2: n === 0 ? null : Math.sqrt(sumSq / n),
    };
  });
}
